### DISCUSSION TITLE
dynamically adjust AMR subcycling to stay within CFL

### DISCUSSION THREAD

**Discussion by BenWibking** [AMReX-Codes/amrex#2806] [Q&A] [answered]
I'm having a problem with stability in very strong shocks/rarefactions when there are several (>4) AMR levels. It looks like the CFL constraint is violated at the finest AMR levels because the CFL-stable timestep has become smaller over a coarse step. Since the timestep for each level is only computed at the beginning of the coarse step, this leads to catastrophic instabilities on the finest levels later in a coarse step.

Is there any way to adjust the subcycling factors dynamically as the coarse step progresses -- an example of what I mean is here: https://github.com/applied-numerical-algorithms-group-lbnl/Chombo_3.2/blob/master/lib/src/AMRTimeDependent/AMR.cpp#L1063.

The equivalent function in AMReX's AMR class doesn't seem to do this, but I want to check if it's implemented somewhere else and I missed it.

**Comment by WeiqunZhang** [comment, marked as answer]
`AmrLevel` has a virtual function
```
    /**                                                                                                                                                                                       
    * \brief Do an integration step on this level.  Returns maximum safe                                                                                                                      
    * time step.  This is a pure virtual function and hence MUST                                                                                                                              
    * be implemented by derived classes.                                                                                                                                                      
    */
    virtual Real advance (Real time,
                          Real dt,
                          int  iteration,
                          int  ncycle) = 0;
```
If you subcyle in that funciton, will that work?

**Reply by BenWibking** [reply]
If I recompute the timestep and adaptively subcycle within advance, I think that should fix the stability issues. But it will probably be less accurate compared to adaptively subcycling in the driver, since the finer levels won't be advanced in between the additional subcycles.

**Reply by WeiqunZhang** [reply]
Suppose you have checkpoints for every step and you restart the simulation with a smaller CFL from the checkpoint right before the instability.  How much small the CFL needs to be?  If it's not much, I would say subcycling inside `advance` is acceptable.

If you have time to implement the adaptively sybcycling, you are welcome to submit a PR.

Also, both Amr::coarseTimeStep and Amr::timestep are virtual.  You could override in your code as well.
